File: phone/src/main/java/com/schloesser/masterthesis/presentation/video/VideoActivity.java
package com.schloesser.masterthesis.presentation.video;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PointF;
import android.media.FaceDetector;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.schloesser.masterthesis.R;

import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;

import static com.schloesser.shared.wifidirect.SharedConstants.FRAME_HEIGHT;
import static com.schloesser.shared.wifidirect.SharedConstants.FRAME_WIDTH;
import static com.schloesser.shared.wifidirect.SharedConstants.HEADER_END;
import static com.schloesser.shared.wifidirect.SharedConstants.HEADER_START;
import static com.schloesser.shared.wifidirect.SharedConstants.SERVERPORT;
import static com.schloesser.shared.wifidirect.SharedConstants.TARGET_FPS;


public class VideoActivity extends AppCompatActivity {
    private ImageView mCameraView;
    public MyClientThread mClient;
    public Thread clientThread;
    public Bitmap mLastFrame;

    private final Handler handler = new MyHandler(this);

    private FaceDetector mFaceDetector = new FaceDetector(FRAME_WIDTH, FRAME_HEIGHT, 10);
    private FaceDetector.Face[] faces = new FaceDetector.Face[10];
    private PointF tmp_point = new PointF();
    private Paint tmp_paint = new Paint();
    private Socket mSocket;
    private DataOutputStream outputStream;

    private Runnable mStatusChecker = new Runnable() {
        @SuppressLint("StaticFieldLeak")
        @Override
        public void run() {
            try {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        if (mLastFrame != null) {

                            Bitmap mutableBitmap = mLastFrame.copy(Bitmap.Config.RGB_565, true);

                            final int face_count = mFaceDetector.findFaces(mLastFrame, faces);

                            new AsyncTask<Void, Void, Void>() {

                                @Override
                                protected Void doInBackground(Void... unused) {
                                    try {
                                        outputStream.writeUTF(HEADER_START);
                                        outputStream.writeInt(face_count);
                                        outputStream.writeUTF(HEADER_END);
                                        outputStream.flush();
                                    } catch (IOException e) {
                                        e.printStackTrace();
                                    }
                                    return null;
                                }
                            }.execute();

                            Canvas canvas = new Canvas(mutableBitmap);

                            for (int i = 0; i < face_count; i++) {
                                FaceDetector.Face face = faces[i];

                                tmp_paint.setColor(Color.RED);
                                tmp_paint.setAlpha(100);

                                face.getMidPoint(tmp_point);
                                canvas.drawCircle(tmp_point.x, tmp_point.y, face.eyesDistance(),
                                        tmp_paint);
                            }

                            mCameraView.setImageBitmap(mutableBitmap);
                        }

                    }
                });
            } finally {
                handler.postDelayed(mStatusChecker, 1000 / TARGET_FPS);
            }
        }
    };

    @SuppressLint("StaticFieldLeak")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_video);
        mCameraView = findViewById(R.id.camera_preview);

        new AsyncTask<Void, Void, Void>() {

            @Override
            protected Void doInBackground(Void... unused) {
                // Background Code
                try {
                    ServerSocket ss = new ServerSocket(SERVERPORT);
                    mSocket = ss.accept();

                    try {
                        outputStream = new DataOutputStream(mSocket.getOutputStream());
                    } catch (IOException e) {
                        e.printStackTrace();
                    }

                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(VideoActivity.this, "Connected to:" + mSocket.getInetAddress().toString(), Toast.LENGTH_LONG).show();
                        }
                    });
                    mClient = new MyClientThread(mSocket, handler);
                    clientThread = new Thread(mClient);
                    clientThread.start();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return null;
            }

        }.execute();
        mStatusChecker.run();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (clientThread != null) clientThread.stop();
    }
}

File: phone/src/main/java/com/schloesser/masterthesis/presentation/video/VideoActivity.kt
package com.schloesser.masterthesis.presentation.video

import android.content.Context
import android.graphics.Bitmap
import android.net.wifi.WifiManager
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.schloesser.masterthesis.R
import com.schloesser.masterthesis.presentation.extension.gone
import com.schloesser.masterthesis.presentation.extension.visible
import com.schloesser.shared.wifidirect.SharedConstants.Companion.HEADER_END
import com.schloesser.shared.wifidirect.SharedConstants.Companion.HEADER_START
import com.schloesser.shared.wifidirect.SharedConstants.Companion.SERVERPORT
import kotlinx.android.synthetic.main.activity_video.*
import org.jetbrains.anko.doAsync
import java.io.DataOutputStream
import java.io.IOException
import java.math.BigInteger
import java.net.InetAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.UnknownHostException
import java.nio.ByteOrder


class VideoActivity : AppCompatActivity(), ProcessFrameTask.Callback {

    companion object {
        private const val TAG = "VideoActivity"
    }

    private var socketThread: Thread? = null
    private var outputStream: DataOutputStream? = null
    private var serverSocket: ServerSocket? = null
    private var socket: Socket? = null

    var lastFrame: Bitmap? = null
        @Synchronized set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_video)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        txvIpAddress.text = "Server address: ${wifiIpAddress()}"
    }

    override fun onResume() {
        super.onResume()
        startServer()
    }

    private fun startServer() {
        initServerSocket()
        processingRunnable.run()
    }

    private fun stopServer() {
        if (socketThread != null) socketThread!!.interrupt()
        if (socket != null) socket!!.close()
        if (serverSocket != null) serverSocket!!.close()
        processFrameTask.stop()
    }

    private fun initServerSocket() {
        doAsync {
            try {
                serverSocket = ServerSocket(SERVERPORT)
                socket = serverSocket!!.accept()

                try {
                    outputStream = DataOutputStream(socket!!.getOutputStream())
                } catch (e: IOException) {
                    e.printStackTrace()
                }

                showStatus("Connected to:" + socket!!.inetAddress.toString())

                socketThread = Thread(ServerSocketThread(socket!!, this@VideoActivity) { _ ->
                    showStatus("Connection lost")
                    imvCameraPreview.setImageBitmap(null)
                    imvFace.setImageBitmap(null)

                    stopServer()
                    startServer()
                })
                socketThread!!.start()

            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private val processFrameTask by lazy {
        ProcessFrameTask(this)
    }

    private val processingRunnable = object : Runnable {
        override fun run() {
            try {
                if (lastFrame != null) {
                    processFrameTask.run(lastFrame!!, this@VideoActivity)
                    lastFrame = null // Set as null to avoid multiple processing
                }
            } finally {
                Handler().post(this)
            }
        }
    }

    override fun onProcessFrameResults(faceCount: Int, emotionLabel: String, frame: Bitmap, processedCenterFace: Bitmap?) {
        sendProcessingResults(faceCount, emotionLabel)
        imvCameraPreview.setImageBitmap(frame)
        imvFace.setImageBitmap(processedCenterFace)
    }

    private fun sendProcessingResults(faceCount: Int, emotionLabel: String) {
        doAsync {
            try {
                outputStream!!.writeUTF(HEADER_START)
                outputStream!!.writeInt(faceCount)
                outputStream!!.writeUTF(emotionLabel)
                outputStream!!.writeUTF(HEADER_END)
                outputStream!!.flush()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    private fun showStatus(message: String) {
        runOnUiThread { Toast.makeText(this@VideoActivity, message, Toast.LENGTH_LONG).show() }
    }

    override fun onPause() {
        super.onPause()
        stopServer()
    }

    protected fun wifiIpAddress(): String? {
        val wifiManager = getSystemService(Context.WIFI_SERVICE) as WifiManager
        var ipAddress = wifiManager.connectionInfo.ipAddress

        // Convert little-endian to big-endianif needed
        if (ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
            ipAddress = Integer.reverseBytes(ipAddress)
        }

        val ipByteArray = BigInteger.valueOf(ipAddress.toLong()).toByteArray()

        var ipAddressString: String?
        try {
            ipAddressString = InetAddress.getByAddress(ipByteArray).getHostAddress()
        } catch (ex: UnknownHostException) {
            Log.e("WIFIIP", "Unable to get host address.")
            ipAddressString = null
        }

        return ipAddressString
    }
}
